Source/Control/adsr.cpp

```diff
@@ -3,79 +3,77 @@
 
 using namespace daisysp;
 
-float Adsr::Tau2Pole(float tau)
+
+void Adsr::Init(float sample_rate, int blockSize)
 {
-    return expf(-1.0f / (tau * sample_rate_));
+    sample_rate_ = sample_rate / blockSize;
+    SetTime(ADSR_SEG_ATTACK,  0.1f);
+    SetTime(ADSR_SEG_DECAY,   0.1f);
+    SetTime(ADSR_SEG_RELEASE, 0.1f);
+    sus_level_  = 0.7f;
+    x_          = 0.0f;
 }
 
-float Adsr::AdsrFilter()
+void Adsr::Retrigger(bool hard)
 {
-    y_ = b_ * x_ + a_ * y_;
-    return y_;
+    mode_ = ADSR_SEG_ATTACK;
+    if(hard)
+        x_ = 0.f;
 }
 
-
-void Adsr::Init(float sample_rate)
+void
+Adsr::SetTime(int seg, float time)
 {
-    seg_time_[ADSR_SEG_ATTACK]  = 0.1f;
-    seg_time_[ADSR_SEG_DECAY]   = 0.1f;
-    sus_                        = 0.7f;
-    seg_time_[ADSR_SEG_RELEASE] = 0.1f;
-    //timer_ = 0;
-    a_           = 0.0f;
-    b_           = 0.0f;
-    x_           = 0.0f;
-    y_           = 0.0f;
-    prev_        = 0.0f;
-    atk_time_    = seg_time_[ADSR_SEG_ATTACK] * sample_rate;
-    sample_rate_ = sample_rate;
+    if(seg_time_[seg] != time)
+    {
+        seg_time_[seg] = time;
+        if(time > 0.f)
+        {
+            const float attackTarget = logf(0.5f);
+            const float decayTarget  = logf(1./M_E);
+            float target = (seg == ADSR_SEG_ATTACK)? attackTarget : decayTarget;
+            seg_D0_  [seg] = 1.f - expf(target / (time * sample_rate_));
+        }
+        else
+            seg_D0_  [seg] = 1.f; // instant change
+    }
 }
 
+
 float Adsr::Process(bool gate)
 {
-    float pole, out;
-    out = 0.0f;
+    float out = 0.0f;
 
     if(gate && mode_ != ADSR_SEG_DECAY)
-    {
         mode_ = ADSR_SEG_ATTACK;
-        //timer_ = 0;
-        pole      = Tau2Pole(seg_time_[ADSR_SEG_ATTACK] * 0.6f);
-        atk_time_ = seg_time_[ADSR_SEG_ATTACK] * sample_rate_;
-        a_        = pole;
-        b_        = 1.0f - pole;
-    }
     else if(!gate && mode_ != ADSR_SEG_IDLE)
-    {
         mode_ = ADSR_SEG_RELEASE;
-        pole  = Tau2Pole(seg_time_[ADSR_SEG_RELEASE]);
-        a_    = pole;
-        b_    = 1.0f - pole;
-    }
-
-    x_    = (int)gate;
-    prev_ = (int)gate;
 
+    float D0(seg_D0_[mode_]);
+    float target = mode_ == ADSR_SEG_DECAY? sus_level_ : - 0.1f;
     switch(mode_)
     {
         case ADSR_SEG_IDLE: out = 0.0f; break;
         case ADSR_SEG_ATTACK:
-            out = AdsrFilter();
-
-            if(out > .99f)
+            x_ += D0 * (2.f - x_);
```

Comment:
This is not a linear attack, right? I guess with the target of 2.0 its not horribly exponential either, but definitely not linear like many people would expect.

I'm wondering if we could make this a parameter when setting the attack time: something like this 
```cpp
void setAttackTime(float timeInS, float shape = 0.0f)
```

Where shape = 0..1 with 
- `shape==0` => `target = 100.0f` for a near-linear response. (time constant calculation would of course have to be adjusted taking into account that we want to reach 1.0 after the attack time independently of the target)
- `shape==1` => `target = 1.0f + someSmallDelta` for a exponential response (where `someSmallDelta` helps to actually reach 1.0, identically to the release phase.

Reply:
Yes, it's a not linear, but less exponential as it was. I like the idea of a shape parameter, that's was really something I hoped to get as feedback. So, yes, I'll put this.

Reply:
Actually, I was thinking about this some more while riding the bike.
If we use a shape parameter to set the target value to something much higher than 1 in order to achieve a "near linear" response, we end up reaching that linear response only when `target` is infinitely large and `D0` is infinitely small. In order to get a nice working range for this shape parameter, we would have to make `target` move along some sort of hyperbola trajectory, right? I'm wondering if we should instead to some sort of blend between an exponential and a linear curve:

```cpp
// linear curve
x_ += someConstant;
// expo curve (target = 1.05f or similar):
x_ += D0 * (target - x_);

// crossfaded version:
x_ += shape * D0 * (target - x_) + (1.0f - shape) * someConstant;

// We can optimize this by pre-calculating two values:
A = shape * D0;
B =  (1.0f - shape) * someConstant;

// Then this would become our update function:
x_ = A * (target - x_) + B_;
```

This crossfaded version should behave much more evenly wrt. the shape parameter. AND we don't have to do any division like we would have to

Reply:
I fired up Grapher, my universal math tool, and I saw, that with shape set to 10, it's almost linear. Mixing linear and exponential stuff looks a bit weird, but could be interesting. 
[LinLog ADSR attack.zip](https://github.com/electro-smith/DaisySP/files/6521017/LinLog.ADSR.attack.zip)
![image](https://user-images.githubusercontent.com/47812706/119103508-8ef16480-ba1b-11eb-894c-8737a172bbbd.png)

Reply:
See possible shapes below.